source-http-ingest/src/server.rs

```diff
@@ -0,0 +1,419 @@
+        };
+
+        let enhanced_doc = match collection.prepare_document(body, request_headers) {
+            Ok(doc) => doc,
+            Err(err) => return Ok(err_response(StatusCode::BAD_REQUEST, err)),
+        };
+
+        let validation_result = collection.validator.validate(None, &enhanced_doc)?;
+        if let Err(validation_err) = validation_result.ok() {
+            return Ok((
+                StatusCode::BAD_REQUEST,
+                Json(
+                    serde_json::json!({"error": "request body failed validation", "validationError": validation_err }),
+                ),
+            ));
+        }
+        let txn = vec![(collection.binding_index, enhanced_doc)];
+        std::mem::drop(handlers_guard);
+
+        self.io.publish(txn).await?;
+        Ok((StatusCode::OK, Json(serde_json::json!({"published": 1 }))))
+    }
+}
+
+const JSON_CONTENT_TYPE: &str = "application/json";
+
+pub fn openapi_spec<'a>(
+    endpoint_config: &EndpointConfig,
+    bindings: &[Binding],
+) -> anyhow::Result<OpenApi> {
+    let failure_schema = schema::ObjectBuilder::new()
+        .property(
+            "error",
+            schema::Object::with_type(schema::SchemaType::String),
+        )
+        .build();
+    let success_schema = schema::ObjectBuilder::new()
+        .property(
+            "published",
+            schema::Object::with_type(schema::SchemaType::Integer),
+        )
+        .build();
+    let mut components = schema::ComponentsBuilder::new()
+        .response(
+            "successResponse",
+            openapi::ResponseBuilder::new().content(
+                JSON_CONTENT_TYPE,
+                openapi::content::ContentBuilder::new()
+                    .example(Some(serde_json::json!({"published": 1})))
+                    .schema(success_schema)
+                    .build(),
+            ),
+        )
+        .response(
+            "failureResponse",
+            openapi::ResponseBuilder::new().content(
+                JSON_CONTENT_TYPE,
+                openapi::content::ContentBuilder::new()
+                    .example(Some(
+                        serde_json::json!({"error": "missing required header 'X-Webhook-Id'"}),
+                    ))
+                    .schema(failure_schema)
+                    .build(),
+            ),
+        );
+    if endpoint_config.require_auth_token.is_some() {
+        let sec = security::SecurityScheme::Http(
+            security::HttpBuilder::new()
+                .scheme(security::HttpAuthScheme::Bearer)
+                .build(),
+        );
+        components = components.security_scheme("bearerAuth", sec);
+    }
+
+    let mut paths = openapi::PathsBuilder::new();
+
+    for binding in bindings.iter() {
+        let url_path = ensure_slash_prefix(binding.resource_path[0].as_str());
+
+        let raw_schema = binding.collection.schema.as_ref().unwrap();
+        let openapi_schema = serde_json::from_str::<openapi::Schema>(raw_schema.get())
```

Comment:
Is it possible for there to be a kind of headers inception here? For example, a schema is inferred on a collection generated from ingested webhooks. Then a user goes to their openapi input and see those headers as inputs and provides them as a manual webhook. But also while doing that headers (maybe the same ones?) are added separately by their browser or whatever.

Reply:
Yeah, and the server will overwrite the user-provided headers with the ones from the request. I just kinda shrugged it off as "too bad there's nothing we can do", but then I just thought of a solution: `readSchema`!

The openapi spec is generated from the `writeSchema` if one is set. So we can just have discover output a `writeSchema` that _doesn't_ include the headers, and a `readSchema` that does include them.

I'm glad you mentioned it, because I think that could be legitimately confusing to users, and having the headers be included in the examples was just noise in the best case. This _will_ still be a little weird with `/_meta/webhookId`, though, since the id has to be required by the `writeSchema`. So users will still see a `webhookId` in the example documents.

Reply:
Oh, nevermind then. The DiscoverResponse only has `documentSchemaJson`. There's no way to return separate read vs write schemas at this point. The only other idea that had come to mind was to have a special case in the server that would try remove those fields from the schema when it generates the openapi schema. That seemed a little too gross to do at this point, but now it's kinda seeming like A, our only option, and B, not any worse than providing a confusing UX. I'll give it a shot and see how bad it is.